### PR TITLE
Bump H2 version and add support for H2v2

### PR DESCRIPTION
**Pull request by ctosin**
H2 v2.x is not compatible with v1.x. After the version bump, some changes to the code were necessary and the legacy mode was enabled. That increases compatibility between versions.

### CODE REVIEW

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java [hunk 1]
```diff
@@ -56,6 +58,10 @@
  * @since 2.0.0
  */
 public class H2Engine extends AbstractDatabaseEngine {
+    /**
+     * Tha max length of a VARCHAR type.
```
Comment (waldyrious) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java:62]:
```suggestion
     * The max length of a VARCHAR type.
```

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java [hunk 2]
```diff
@@ -355,4 +355,20 @@ public String translateTrue() {
     public String translateFalse() {
         return "FALSE";
     }
+
+    /**
+     * In H2 2.x, checking a boolean field against a boolean represented as a string causes an error.
+     * This code checks if the string is a boolean and process it as a boolean, not as a string.
```
Comment (waldyrious) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java:361]:
```suggestion
     * This code checks if the string is a boolean and processes it as a boolean, not as a string.
```

File: src/test/resources/connections.properties [hunk 3]
```diff
@@ -5,14 +5,18 @@
 # the ports specified in the JDBC URLs are the default ones and are used by the Timeouts test
 ###############################################################################################
 
+# -=- H2 LEGACY MODE -=-
+# The version 2.x of H2 is not compatible with the version 1.x. When the version was bumped, the H2 legacy
+# mode was turned on to keep some of the backward compatibility with 1.x version. For more information, see:
```
Comment (waldyrious) [inline src/test/resources/connections.properties:10]:
```suggestion
# Version 2.x of H2 is not compatible with version 1.x. When the version was bumped to 2.x,
# the H2 legacy mode was turned on to keep some backward compatibility with the 1.x version.
# For more information, see:
```

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java [hunk 4]
```diff
@@ -56,6 +58,10 @@
  * @since 2.0.0
  */
 public class H2Engine extends AbstractDatabaseEngine {
+    /**
+     * The max length of a VARCHAR type.
+     */
+    private static final int MAX_VARCHAR_LENGTH = 1048576;
```
Comment (francisco-polaco) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java:64]: Question: Why this value?

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java [hunk 5]
```diff
@@ -355,4 +355,20 @@ public String translateTrue() {
...
+
+    /**
+     * In H2 2.x, checking a boolean field against a boolean represented as a string causes an error.
+     * This code checks if the string is a boolean and processes it as a boolean, not as a string.
+     */
+    @Override
+    public String translate(final K k) {
+       if (k.getConstant() instanceof String) {
+            String s = (String) k.getConstant();
+            if (s.equalsIgnoreCase("TRUE") || s.equalsIgnoreCase("FALSE")) {
```
Comment (francisco-polaco) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java:367]: Question: Does `"1"` and `"0"` also represent boolean values?

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java (hunk 4 above)
Comment (ctosin) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java:64]: That value comes from the [documentation](http://www.h2database.com/html/datatypes.html#:~:text=CHARACTER%20VARYING,-%7B%20%7B%20CHARACTER%20%7C%20CHAR&text=create%20a%20quote.-,The%20allowed%20length%20is%20from%201%20to%201048576%20characters.,when%20using%20this%20data%20type.)

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java (hunk 5 above)
Comment (francisco-polaco) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java:367]: Makes sense. I was confused, because Oracle uses `1` and `0` to represent Booleans.

**Review by francisco-polaco — APPROVED** [review]
LGTM! :muscle: 

Please, add the link on the javadoc as suggested :)

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java (hunk 4 above)
Comment (francisco-polaco) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java:64]: Alright! I would add that link on the javadoc for reference :)

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java [hunk 6]
```diff
@@ -355,4 +355,20 @@ public String translateTrue() {
...
+                return Boolean.parseBoolean(s) ? translateTrue() : translateFalse();
+            }
+        }
```
Comment (jmf-tls) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java:370]: what if I have a string that happens to be "TRUE" but I really mean it as a string not a boolean??
instead of "TRUE", it's going to write in the DB "1"??

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java [hunk 7]
```diff
@@ -147,7 +153,16 @@ protected int entityToPreparedStatement(final DbEntity entity, final PreparedSta
 
                 switch (column.getDbColumnType()) {
                     case BLOB:
-                        ps.setBytes(i, objectToArray(val));
+                        if (val == null) {
+                            ps.setNull(i, Types.BLOB);
+                            break;
+                        }
+
+                        if (val instanceof Serializable) {
```
Comment (artpdr) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java:162]: I would use an `else if` here and remove the break from the block above...

File: src/test/resources/connections.properties [hunk 8]
```diff
@@ -5,14 +5,18 @@
...
 ###############################################################################################
 
+# -=- H2 LEGACY MODE -=-
+# Version 2.x of H2 is not compatible with version 1.x. When the version was bumped to 2.x,
+# the H2 legacy mode was turned on to keep some backward compatibility with the 1.x version.
+# For more information, see: http://www.h2database.com/html/features.html.
 
 h2.engine=com.feedzai.commons.sql.abstraction.engine.impl.H2Engine
-h2.jdbc=jdbc:h2:./target/pdb;AUTO_SERVER=FALSE
+h2.jdbc=jdbc:h2:./target/pdb;AUTO_SERVER=FALSE;MODE=LEGACY
```
Comment (jmf-tls) [inline src/test/resources/connections.properties:14]: This doesn't seem very bad for now, but in which ways is v2.x not compatible with v1.x?
Would it be too much trouble creating an implementation of H2Enginev2?

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java [hunk 9]
```diff
@@ -686,4 +701,20 @@ protected ResultIterator createResultIterator(PreparedStatement ps) throws Datab
     protected QueryExceptionHandler getQueryExceptionHandler() {
         return H2_QUERY_EXCEPTION_HANDLER;
     }
+
+    @Override
+    protected void setParameterValues(PreparedStatement ps, int index, Object param) throws SQLException {
```
Comment (artpdr) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java:707]: Can you make these final?

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java [hunk 10]
```diff
@@ -686,4 +701,20 @@ protected ResultIterator createResultIterator(PreparedStatement ps) throws Datab
...
+        if (param instanceof String) {
+            String paramStr = (String) param;
```
Comment (artpdr) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java:709]: final

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java [hunk 11]
```diff
@@ -355,4 +355,20 @@ public String translateTrue() {
...
     }
+
+    /**
+     * In H2 2.x, checking a boolean field against a boolean represented as a string causes an error.
+     * This code checks if the string is a boolean and process it as a boolean, not as a string.
+     */
+    @Override
+    public String translate(final K k) {
+       if (k.getConstant() instanceof String) {
+            String s = (String) k.getConstant();
```
Comment (artpdr) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java:366]: final

File: src/test/resources/connections.properties [hunk 12]
```diff
@@ -5,14 +5,18 @@
...
 ###############################################################################################
 
+# -=- H2 LEGACY MODE -=-
+# The version 2.x of H2 is not compatible with the version 1.x. When the version was bumped, the H2 legacy
+# mode was turned on to keep some of the backward compatibility with 1.x version. For more information, see:
+# http://www.h2database.com/html/features.html.
 
 h2.engine=com.feedzai.commons.sql.abstraction.engine.impl.H2Engine
-h2.jdbc=jdbc:h2:./target/pdb;AUTO_SERVER=FALSE
+h2.jdbc=jdbc:h2:./target/pdb;AUTO_SERVER=FALSE;MODE=LEGACY
```
Comment (artpdr) [inline src/test/resources/connections.properties:14]: Is it necessary for everyone using H2 on pdb to add this option?

**Comment by artpdr** [comment]
On the commit message use just `Bump H2 version and use legacy mode for tests`

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java (hunk 7 above)
Comment (jmf-tls) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java:162]: I don't think we even need to have the special case for null here....
Check H2 driver internals, it seems to handle _null_ just fine

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java (hunk 6 above)
Comment (ctosin) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java:370]: The idea of this method is to consider "TRUE" or "FALSE" (strings) as TRUE or FALSE (booleans). Basically it removes the quotes. In a situation like this, the code would throw an error, because the unquoted value could not be stored in a string field.

I am aware that this implementation is not perfect. But CM is used to use booleans as strings in several select expressions, which is not allowed anymore in H2. So without this method, CM would have several breaking changes after the bump. So it is trade-off. In my opinion, it would be better to support that directly in PDB and from now on always use boolean comparisons as booleans, not strings.

File: src/test/resources/connections.properties (hunk 8 above)
Comment (ctosin) [reply src/test/resources/connections.properties:14]: The migration changes are documented here: https://www.h2database.com/html/migration-to-v2.html

I don't think that is necessary. I did tests with CM and Genome so far and none of them need the legacy mode. The legacy mode here in PDB is mainly because there are several tests that would fail without that (mainly because the way that H2 handled the ID counter in auto-increment columns, which changed in the new version). Fixing these tests is not trivial and that is why I opted to use the legacy mode.

File: src/test/resources/connections.properties (hunk 12 above)
Comment (ctosin) [reply src/test/resources/connections.properties:14]: No. The legacy mode is for testing purposes. PDB clients can choose to use it or not.

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java (hunk 7 above)
Comment (ctosin) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java]: The problem here is not the driver, but the `entry` map. This map may not have the key, so it is necessary to check for nulls.

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java (hunk 9 above)
Comment (ctosin) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java]: Yep!

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java (hunk 10 above)
Comment (ctosin) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java]: Yep!

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java (hunk 11 above)
Comment (ctosin) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java]: Yep!

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java (hunk 6 above)
Comment (ctosin) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Translator.java]: Ok, I see your point and I agree. I will remove that and clients of PDB will have to use boolean as boolean values.

File: src/test/resources/connections.properties (hunk 12 above)
Comment (artpdr) [reply src/test/resources/connections.properties:14]: I'm asking this because now we just test with legacy mode on. We aren't testing when the legacy mode is off which should happen in the majority of clients from this lib.

File: README.md [hunk 13]
```diff
@@ -37,6 +37,11 @@ DatabaseEngine methods `#getPSResultSet(final String name)`, `#getPSIterator(fin
 (final String name, final int fetchSize)`, since 2.7.0, throw an extra `ConnectionResetException` when the database
 connection is lost and recovered.
 
+H2 2.x is not backward compatible with H2 1.x. Two changes worth some comments:
```
Comment (artpdr) [inline README.md:40]: I think this will require a release of PDB 2.9.0. So please put this text in a new header (## Changes from 2.9.0) after ## Changes from 2.8.0

Comment (ctosin) [reply README.md]: :+1:

File: src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/h2/H2NotLegacyTest.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+import static com.feedzai.commons.sql.abstraction.dml.dialect.SqlBuilder.dbEntity;
+import static com.feedzai.commons.sql.abstraction.dml.dialect.SqlBuilder.entry;
+import static com.feedzai.commons.sql.abstraction.util.StringUtils.quotize;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests H2 behavior in not legacy mode.
+ *
+ * @author Carlos Tosin (carlos.tosin@feedzai.com)
+ * @since ```feedzai.next.release```
```
Comment (artpdr) [inline src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/h2/H2NotLegacyTest.java:43]: Don't use this in pdb. You will need to put the version hardcoded - 2.9.0

File: src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/EngineGeneralTest.java [hunk 15]
```diff
@@ -168,7 +168,9 @@
 
     @Parameterized.Parameters
     public static Collection<DatabaseConfiguration> data() throws Exception {
-        return DatabaseTestUtil.loadConfigurations();
+        return DatabaseTestUtil.loadConfigurations(
+                "h2Legacy", "h2RemoteLegacy", "mysql", "sqlserver", "postgresql", "cockroach", "oracle", "db2"
```
Comment (artpdr) [inline src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/EngineGeneralTest.java:172]: Why this change?

**Review by artpdr — COMMENTED** [review]
I'm not comfortable with what we are doing with tests. We are testing just some things in the non legacy version. Some things that were tested before like in `BatchConnectionRetryTest` aren't being tested anymore.

File: src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/EngineGeneralTest.java (hunk 15 above)
Comment (ctosin) [reply src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/EngineGeneralTest.java:172]: Because this test class needs do load configurations of all of the database types, except non-legacy H2 (the specific non-legacy H2 tests are in another class - `H2NotLegacyTest`). So it is necessary to specify each one explicitly.

File: src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/h2/H2NotLegacyTest.java (hunk 14 above)
Comment (ctosin) [reply src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/h2/H2NotLegacyTest.java:43]: 👍

File: .github/workflows/test-h2v2.yml [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    name: Test H22 PDB Java ${{ matrix.java-version }}
```
Comment (artpdr) [inline .github/workflows/test-h2v2.yml:15]: H2v2

File: README.md [hunk 17]
```diff
@@ -46,6 +46,11 @@ connection is lost and recovered.
 ## Changes from 2.8.0
 * It now uses Guava 25.1-jre version, which might require the client to also upgrade it to match the same version
 
+## Changes from 2.8.10
+* H2 version upgraded to 2.1.210
+* `H2Engine` is now deprecated and uses the H2 legacy mode (for more information see: http://www.h2database.com/html/features.html)
```
Comment (artpdr) [inline README.md:51]: ...the H2 v2 legacy mode if this engine is used with the h2 v2 driver...

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/DatabaseEngineDriver.java [hunk 18]
```diff
@@ -38,6 +38,20 @@ public String connectionStringFor(String hostname, String database, int port) {
...
+     * The H2v2 Database vendor.
+     */
+    H2V2("H2V2", "com.feedzai.commons.sql.abstraction.engine.impl.H2EngineV2", "org.h2.Driver", 0) {
+        @Override
+        public String connectionStringFor(String hostname, String database) {
+            return String.format("jdbc:h2:%s/%s", hostname, database);
+        }
+
+        @Override
+        public String connectionStringFor(String hostname, String database, int port) {
```
Comment (artpdr) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/DatabaseEngineDriver.java:51]: We don't care about the port?

**Review by artpdr — APPROVED** [review]

File: README.md (hunk 17 above)
Comment (ctosin) [reply README.md]: :+1:

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/DatabaseEngineDriver.java (hunk 18 above)
Comment (ctosin) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/DatabaseEngineDriver.java:51]: I just followed the convention that was already being used by `H2`. I guess that's because this code relates to the embedded mode of H2, which does not use a port.

File: .github/workflows/test-h2v2.yml (hunk 16 above)
Comment (ctosin) [reply .github/workflows/test-h2v2.yml]: :+1:

**Review by jmf-tls — APPROVED** [review]
